### PR TITLE
similar_names fixes

### PR DESCRIPTION
**Pull request by oli-obk**
* don't lint at the use-site of bad struct field bindings if they're shorthand
    * see #899
* properly lint function argument patterns in similar_names
    * I used `visit::walk_pat` instead of visiting the pattern directly, this mean only sub-patterns got linted against

### CODE REVIEW

File: tests/compile-fail/non_expressive_names2.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#![feature(plugin)]
```
Comment (llogiq) [inline tests/compile-fail/non_expressive_names2.rs]: Is there a reason why this is in its own file?

Comment (oli-obk) [inline tests/compile-fail/non_expressive_names2.rs]: stuff wasn't working and compiletest doesn't produce nice error messages. I'll merge it back in the normal file

**Comment by llogiq** [comment]
r=me modulo compiletest nit.

**Comment by oli-obk** [comment]
done

**Pull request merged**